lib/sycamore/sycamore/connectors/neo4j/neo4j_writer.py

```diff
@@ -0,0 +1,343 @@
+            return doc.data
+
+        def accumulate_row(headers, row):
+            data = extract_nodes(row)
+            if "label" not in data:
+                return headers
+            node_key = data["label"]
+            if headers["nodes"].get(node_key, None) is None:
+                headers["nodes"][node_key] = dict()
+            #### add all required keys ####
+            #### ALL DOCS HAVE uuid:ID ####
+            headers["nodes"][node_key]["uuid:ID"] = True
+            #### IF KEYS EXIST IN DATA ####
+            for key in ["type", "bbox", "text_representation"]:
+                if key in data:
+                    headers["nodes"][node_key][key] = True
+            #### add all keys from properties ####
+            for key in data["properties"].keys():
+                headers["nodes"][node_key][key] = True
+
+            for key, value in data["relationships"].items():
+                rel_key = value["START_LABEL"] + "_" + value["END_LABEL"]
+                if headers["relationships"].get(rel_key, None) is None:
+                    headers["relationships"][rel_key] = dict()
+                #### add all required keys ####
+                for key in ["uuid:ID", ":START_ID", ":END_ID", ":TYPE"]:
+                    headers["relationships"][rel_key][key] = True
+                #### add all keys from properties ####
+                for key in value["properties"].keys():
+                    headers["relationships"][rel_key][key] = True
+            return headers
+
+        def merge(headers1, headers2):
+            #### merge nodes together ####
+            for key, values in headers2["nodes"].items():
+                if headers1["nodes"].get(key, None) is None:
+                    headers1["nodes"][key] = values
+                else:
+                    for value in values:
+                        headers1["nodes"][key][value] = True
+            #### merge relationships together ####
+            for key, values in headers2["relationships"].items():
+                if headers1["relationships"].get(key, None) is None:
+                    headers1["relationships"][key] = values
+                else:
+                    for value in values:
+                        headers1["relationships"][key][value] = True
+            return headers1
+
+        def finalize(nodes):
+            return nodes
+
+        aggregation = AggregateFn(
+            init=lambda group_key: dict(nodes=dict(), relationships=dict()),
+            accumulate_row=accumulate_row,
+            merge=merge,
+            finalize=finalize,
+            name="output",
+        )
+
+        return self._dataset.aggregate(aggregation)["output"]
+
+    def _write_nodes_csv_headers(self, nodes):
+        for node_label, node_columns in nodes.items():
+            csv_path = os.path.expanduser(f"{self._import_dir}/sycamore/nodes/{node_label}.csv")
+            headers = sorted(set(column for column in node_columns.keys()))
```

Comment:
its easiest to make this a list here. We need to mantain consistent order of the headers otherwise it will right values to the wrong column. sorted(set()) solves this but its slow, we can just use a list since all the columns will have unique values after the aggregate.

switched to using list.